### DISCUSSION TITLE
Support for non-complient OIDC IDP, that support OAuth2

### DISCUSSION THREAD

**Discussion by Herdo** [zitadel/zitadel#4125] [Ideas]
Other IAM solutions offer the ability to register IDPs that implement the OAuth2 standard, but not OIDC.
One example would be Discord, who offer OAuth2 endpoints, but no discovery document for OIDC.

Could this be added as third option, next to OIDC Providers and JWT Providers?

**Comment by Herdo** [comment]
This usually involves manual mapping of endpoints as well as claims.
Other IAM solutions like Keycloak, Auth0 and FusionAuth provide this option.

[FusionAuth](https://fusionauth.io/docs/v1/tech/identity-providers/openid-connect/discord) and [Keycloak](https://www.keycloak.org/docs/latest/server_admin/#identity-provider-operations) have docs regarding this:
![grafik](https://user-images.githubusercontent.com/651191/183072385-1982e957-7e1b-4bf6-a7c2-513b6016534d.png)

**Reply by mffap** [reply]
Thank you for sharing the config details. We'll have a look at it when we start https://github.com/zitadel/zitadel/issues/2998

**Reply by livio-a** [reply]
We will create templates for different IDPs / Social Logins (https://github.com/zitadel/zitadel/issues/2998), this includes `Generic OAuth 2.0` which will look something like the ones mentioned.

**Reply by mffap** [reply]
@Herdo With the latest version of ZITADEL we published a set of IDP templates, including the Generic OAuth 2.0 provider mentioned by @livio-a - Can you try it out, if that works for you with Discord?

<img width="1242" alt="Screenshot 2023-03-22 at 16 26 02" src="https://user-images.githubusercontent.com/1366906/226935652-3ff7a5bf-bbaf-43ca-8138-c5b843faec3f.png">

**Comment by fuomag9** [comment]
I don't know if gitlab is OIDC compliant (but it does publish a .well-known file), but configuring it right now results in:

```
oauth2: cannot fetch token: 400 Bad Request Response: {"error":"invalid_request","error_description":"Missing required parameter: code."}
```

When trying to sign in or register

**Reply by adlerhurst** [reply]
We figured out the problem. Gitlab as idp is not fully oidc compliant. We currently send the optional `prompt`-parameter which Gitlab does not handle and does not ignore. 

Example:
currently sent by ZITADEL: `https://gitlab.com/oauth/authorize?client_id=<APPLICATION_ID>&prompt=select_account&redirect_uri=https%3A%2F%2F<MY_INSTANCE>.zitadel.cloud%2Fui%2Flogin%2Flogin%2Fexternalidp%2Fcallback&response_type=code&scope=openid+profile+email&state=<STATE>`

without `prompt=select_account` it would work

e.g.: `https://gitlab.com/oauth/authorize?client_id=<APPLICATION_ID>&redirect_uri=https%3A%2F%2F<MY_INSTANCE>.zitadel.cloud%2Fui%2Flogin%2Flogin%2Fexternalidp%2Fcallback&response_type=code&scope=openid+profile+email&state=<STATE>`

**Reply by adlerhurst** [reply]
ps. I would not play with it too much. It looks like I just broke my account 🥳

**Reply by fuomag9** [reply]
As I had imagined it was indeed not completely OIDC compliant (this will be very useful on my other projects to know)
Thank you for the nice info (and I hope it can be integrated soon!)

**Reply by CMiksche** [reply]
There is a Issue at GitLab about this: https://gitlab.com/gitlab-org/gitlab/-/issues/377368

**Reply by xXentaOps** [reply]
FYI to save people a click: that issue at GitLab is still open 2.5y later (according to [a comment on 25Feb2025](https://gitlab.com/gitlab-org/gitlab/-/issues/377368#note_2365534567))
